### PR TITLE
Fixes Libresonic/libresonic#451

### PR DESCRIPTION
**Pull request by akpdw**
This replaces the existing FolderBasedContentDirectory with a new
DispatchingContentDirectory.  The new system then dispatches requests to
subclasses of UpnpContentProcessor, which then handle the requests as
appropriate.

This version also implements a new searchByName method in SearchService,
which is used to provide artist, album, and track search functionality
for the upnp service.

Guidelines for Contributing
---------------------------

Libresonic development is a community project, and contributions are welcomed. Here are a few guidelines you should follow before submitting:

  1.  **License Acceptance** All contributions must be licensed as [GNU GPLv3](https://github.com/Libresonic/libresonic/blob/develop/LICENSE.txt) to be accepted. Use [`git commit --signoff`](https://jk.gs/git-commit.html) to acknowledge this.
  2.  **No Breakage** New features or changes to existing ones must not degrade the user experience. This means do not introduce bugs, remove functionality, or make large changes to existing themes/UI without prior discussion in an Issue.
  3.  **Coding standards** Language best-practices should be followed, comment generously, and avoid "clever" algorithms. Refactoring existing messes is great, but watch out for breakage.
  4.  **Be bold!** Without contributions, this project will vanish. If you just want to help out, try [submiting a patch](https://github.com/Libresonic/libresonic/issues?q=is%3Aissue+is%3Aopen+label%3Apatches-welcome) for an open Issue.
  5.  **Stay relevant** Issues or commentary that is off-topic or tangential to Libresonic development is subject to moderation. Questions should be focused on improving documentation to solve a problem. Visit [Reddit](https://www.reddit.com/r/libresonic) or [IRC](http://webchat.freenode.net?channels=%23libresonic) for community discussion.

### CODE REVIEW

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/DispatchingContentDirectory.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+import java.util.Arrays;
+
+/**
+ * @author Allen Petersen
+ * @author Sindre Mehus
+ * @version $Id$
+ */
+public class DispatchingContentDirectory extends LibresonicContentDirectory {
+
+    public static final Logger LOG = LoggerFactory.getLogger(FolderBasedContentDirectory.class);
```
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/DispatchingContentDirectory.java]: Logger class should be `DispatchingContentDirectory`

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/DispatchingContentDirectory.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            String[] splitId = objectId.split(SEPARATOR);
+            String browseRoot = splitId[0];
+            String itemId = splitId.length == 1 ? null : splitId[1];
+
+            UpnpContentProcessor processor = findProcessor(browseRoot);
+            if (processor == null) {
+                // if it's null then assume it's a file, and that the id
+                // is all that's there.
+                itemId = browseRoot;
+                browseRoot = "folder";
```
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/DispatchingContentDirectory.java]: Why is this needed? Looks like browseRoot isnt used after this point.

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/AlbumUpnpProcessor.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        // TODO: correct artist?
+        if (album.getArtist() != null) {
+            container.setArtists(getAlbumArtists(album.getArtist()));
+        }
+        return container;
+    }
+
+    public List<Album> getAllItems() {
+        List<MusicFolder> allFolders = getDispatchingContentDirectory().getSettingsService().getAllMusicFolders();
+        return getAlbumDao().getAlphabetialAlbums(0, 0, false, allFolders);
```
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/AlbumUpnpProcessor.java]: Album browsing didnt work for me and I think its because this only ever returns zero albums.

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    public  List<Album> getChildren(Artist artist) {
+        List<MusicFolder> allFolders = getDispatcher().getSettingsService().getAllMusicFolders();
+        List<Album> allAlbums = getAlbumProcessor().getAlbumDao().getAlbumsForArtist(artist.getName(), allFolders);
+        if (allAlbums.size() > 1) {
+            Album viewAll = new Album();
+            viewAll.setName("- All Albums -");
+            viewAll.setId(-1);
+            viewAll.setComment(AlbumUpnpProcessor.ALL_BY_ARTIST + "_" + artist.getId());
+            allAlbums.add(0, viewAll);
+        }
```
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java]: What is the point of this? Wouldn't someone just go back to all artists if they wanted to. Seems strange to add custom navigation.

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/RecentAlbumUpnpProcessor.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    public List<Album> getAllItems() {
+        List<MusicFolder> allFolders = getDispatchingContentDirectory().getSettingsService().getAllMusicFolders();
+        List<Album> recentAlbums = getAlbumDao().getNewestAlbums(0, RECENT_COUNT, allFolders);
+        if (recentAlbums.size() > 1) {
+            Album viewAll = new Album();
+            viewAll.setName("- All Albums -");
+            viewAll.setId(-1);
+            viewAll.setComment(AlbumUpnpProcessor.ALL_RECENT);
+            recentAlbums.add(0, viewAll);
+        }
```
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/RecentAlbumUpnpProcessor.java]: Same as in the artist processor, why do we need this?

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java (hunk 4 above)
Comment (akpdw) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java]: Browsing by artist goes (list all artists) - (list albums by selected artist) - (list songs on selected album).  The '  - All Albums - ' option allows for the third step to be (list all songs on all albums by that artist).

I added a comment to the code to try to make it clearer.

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/RecentAlbumUpnpProcessor.java (hunk 5 above)
Comment (akpdw) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/RecentAlbumUpnpProcessor.java]: Same as above, in this case allowing you to select all songs on all recent albums.

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/AlbumUpnpProcessor.java (hunk 3 above)
Comment (akpdw) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/AlbumUpnpProcessor.java]: Huh.  I think that album list is populated from ID3 tags?  My music collection loaded in, if not all, then at least most of my albums, but I didn't check all that closely.

**Comment by akpdw** [comment]
I also added in the broadcast search capabilities change.  The two clients I've been using apparently don't pay attention to that at all.  DLNA is such a wonderfully strictly-supported spec. :)

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java (hunk 4 above)
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java]: Sure makes sense. But it seems weird we are adding it as a `org.libresonic.player.domain.Album`. Really we should be adding this at the `org.fourthline.cling.support.model.container.Container` level which would also probably fix the display to look like a folder on upnp clients.

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/AlbumUpnpProcessor.java (hunk 3 above)
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/AlbumUpnpProcessor.java]: Okay I think I figured out the issue. For hsql, it returns all records in the case that you send limit 0 which is weird behavior possibly related to https://sourceforge.net/p/hsqldb/bugs/1100/. However, being on postgres this is actually returning zero rows for me. The naive fix here is to change / add a DAO level function to return all albums in alpha order. However, the better change IMO would be to utilize the pagination options that were passed in at `org.libresonic.player.service.upnp.UpnpContentProcessor#browseRoot` rather than doing the in memory filtering as seen at `org/libresonic/player/service/upnp/UpnpContentProcessor.java:81`

File: libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java (hunk 4 above)
Comment (akpdw) [inline libresonic-main/src/main/java/org/libresonic/player/service/upnp/ArtistUpnpProcessor.java]: It gets added as a org.libresonic.player.domain.Album to go along with the actual Albums that the Artist has.  Those then are passed on to addChild(), which in turn calls getAlbumProcessor().createContainer(album), and adds it to the response as a  org.fourthline.cling.support.model.container.MusicAlbum.  So it does get added as a Container eventually.

You could argue that since this isn't a proper MusicAlbum that it (both the all by artist and all recent albums) is really more of a playlist, and should be represented by a org.fourthline.cling.support.model.container.PlaylistContainer instead.  I don't know if that would gain anything on a practical level, but it could be handled that way instead.

**Comment by muff1nman** [comment]
Good work here, merged this into the develop branch. Still need to test since the album changes but any further tweaks can be done in the develop branch.
